### PR TITLE
Enabling bots to make bids

### LINKED ISSUE #3: Allowing bots to make bids
When the feature is complete, bots will be able to compete with the player in the bidding process. The initial algorithm for deciding whether to play should just pick whether to go one PlayMode lower or skip a turn with 50% probability. Making a heuristic for the bots will be done later.

### PR DESCRIPTION
**Pull request by urosgluscevic**
Completed the feature of letting bid against the player:
- in the PlayButtonView, bots can make bids (currently done randomly, see JavaDoc)
- bots can pick a king from the talon (currently done randomly, check JavaDoc)
- bots can drop the appropriate number of cards from their hand (currently done by dropping the highest-point legal cards)
- the game goes on as normal

Implementation note:
- I have made interfaces for various bot rules, which define how the bots decide whether to play for a lower bid, which king to pick and what cards to put down. The current implementations are either random or greedy, but additional implementations can be made and combined and tested out in different ways, hence the interfaces

Issues:
- In Tarok, if all four players initially skip, the game goes into a "negative" mode, which we have not yet implemented.
- as a result, the human player is not allowed to skip on their first go, making sure that at least someone makes a bid, and that game does not go into the negative mode
- this is a temporary solution, which will be replaced once we implement the negative gamemode

Note:
- please pull the branch onto your local machine and manually test the app, a lot. Also, look at the code and suggest potential refactorings, and also try see if there are any obvious issues with the code

### CODE REVIEW

File: app/src/main/java/com/example/tarok/views/PlayButtonsView.java [hunk 1]
```diff
@@ -26,15 +60,234 @@ public PlayButtonsView(MainActivity mainActivity) {
...
+                }
+            }
+        });
+        playThree.setOnClickListener(view -> {
+            currentLowestBid = 0;
+            playThreeLabel.setText("PLAYER 1");
+            enableAllButtons(false);
+            displayLatestBid();
+            skips = 0;
+            makeBotBids(0);
```
Comment (GabrielTertelici) [inline app/src/main/java/com/example/tarok/views/PlayButtonsView.java]: We should replace all of these with a ClickListener class which takes currentLowestBid as a parameter.

**Review by GabrielTertelici — CHANGES_REQUESTED** [review]
Some things I have noticed:
-We should combine the bot rules regarding what to play for, what king to pick and what to drop
-The Card comparator class should have a better name or a comment saying what it is comparing and what is the resulting order
-We should probably rethink the handleTeammatesOfBots() method in GameStage since it is ridiculous at the moment
-In the play buttons view there is a lot of duplication with the play buttons click listeners (see comment)

File: app/src/main/java/com/example/tarok/views/PlayButtonsView.java (hunk 1 above)
Comment (urosgluscevic) [reply app/src/main/java/com/example/tarok/views/PlayButtonsView.java]: Definitely, the code duplication is horrible here. However, makeBotsBid(int botId) handles some UI stuff from the PlayButtonsView (colours the buttons), so we'll have to refactor that as well to make this possible

**Comment by urosgluscevic** [comment]
Thank you, will look into it. Maybe you could handle the CardComparator because I just copied your code to prevent duplication, and I don't quite get how it works?

**Comment by GabrielTertelici** [comment]
Another bug: the points tally at the end of the game is wrong. If the player is not chosen/does not play, the points sums at the end will be swapped

**Comment by urosgluscevic** [comment]
> Some things I have noticed: 
-We should combine the bot rules regarding what to play for, what king to pick and what to drop 

This is now done in the BotTalonStageRuleManager class. This class in instantiated in the TalonStage, where it is used, and then the same instance is passed to the TalonCardsView and PlayButtonView where it is needed later. It is basically kind of like a Singleton in Spring (back in OOPP), but I could not find the appropriate annotation, nor do we have a nice tool to do dependency injection for us, so I did it manually. 

Please check it out and let me know if we can tick this point off. Thanks

**Comment by urosgluscevic** [comment]
> Some things I have noticed: 
 -In the play buttons view there is a lot of duplication with the play buttons click listeners (see comment)

I have made a BidClickListener class which handles all the clickListeners of all the buttons. This required making some additional getters and setters for the PlayButtonView class, but overall I think the code is much cleaner now. Please let me know what you think.

P.S. I was thinking of extracting the makeBotsBid(int bidderId) method from the PlayButtonsView class into a different class, because the PlayButtonsView should only really have to deal with UI in my opinion. What do you think?

**Comment by urosgluscevic** [comment]
> Some things I have noticed:
 -We should probably rethink the handleTeammatesOfBots() method in GameStage since it is ridiculous at the moment

I have made the BotTeammateHandler class, where I re-implemented the handleTeamMatesOfBots() method from GameStage to make it more comprehensible. I also made tests for every combination of player and teammate. I had to add some stuff to build.gradle, please check if I broke anything there

**Comment by urosgluscevic** [comment]
> P.S. I was thinking of extracting the makeBotsBid(int bidderId) method from the PlayButtonsView class into a different class, because the PlayButtonsView should only really have to deal with UI in my opinion. What do you think?

Did this with the BotBiddingProcess class, which communicates directly with the new BidClickListener. Now, the PlayButtonView only handles the UI, and these two new classes handle the logic and the listeners

**Comment by urosgluscevic** [comment]
> Another bug: the points tally at the end of the game is wrong. If the player is not chosen/does not play, the points sums at the end will be swapped

I tried to replicate this bug but I was not able to. If the human plays or is picked, I pass pointsTeam1 as the first parameter, otherwise as the second, so for me, my (human player) points were always show on top in the tally.

**Review by GabrielTertelici — APPROVED** [review]
Everything looks much cleaner, and the bug regarding the wrong points being displayed has also been fixed. This is ready for merging

**Pull request merged**